### PR TITLE
ExtSvcAuth: Clean up orphaned external services on start up

### PR DESCRIPTION
**Pull request by gamab**
**What is this feature?**

This PR adds a new 1-time background job that cleans up orphaned external services on start up.

Orphaned external services can exist when their associated plugins have been removed through the CLI:
```
 ./grafana/bin/grafana-cli plugins remove plugin-to-remove
```

**Which issue(s) does this PR fix?**:

Fixes #https://github.com/grafana/grafana-enterprise/issues/5989

**Special notes for your reviewer:**

How to test this:
- clone the https://github.com/grafana/grafana-plugin-examples repository
- build and install the `app-with-service-account` and `app-with-on-behalf-of-auth` plugins:
  ```bash
  cd grafana-plugin-examples/examples/app-with-service-account/
  yarn build && mage -v
  cd ../app-with-on-behalf-of-auth
  yarn build && mage -v
  ln -sfn PATH_TO/grafana-plugin-examples/examples/app-with-on-behalf-of-auth/dist/ PATH_TO/grafana/data/plugins/grafana-appwithonbehalfofauth-app
  ln -sfn PATH_TO/grafana-plugin-examples/examples/app-with-service-account/dist/ PATH_TO/grafana/data/plugins/grafana-appwithserviceaccount-app
   ```
- Edit the config ini file with feature flags:
  ```ini
  [feature_toggles]
  enable = externalServiceAccounts externalServiceAuth
  ```
- Run Grafana. Navigate to `plugins` and check both plugins are there (you can enable them). Navigate to `serviceaccounts` and check that both plugins' service accounts are there.
- Remove one plugin:
  ```bash
  ./grafana/bin/grafana-cli plugins remove grafana-appwithserviceaccount-app
  # Removing plugin: grafana-appwithserviceaccount-app
  # Please restart Grafana after installing or removing plugins. Refer to Grafana documentation for instructions if necessary.
  ```
- Restart grafana and check that the service account for `grafana-appwithserviceaccount-app` has been deleted.
- You should also see the following logs:
  ```log
  logger=extsvcauth.registry t=2023-11-14T17:02:35.20268547+01:00 level=info msg="Detected removed External Service" service=grafana-appwithserviceaccount-app provider=ServiceAccounts
  logger=serviceauth.extsvcaccounts t=2023-11-14T17:02:35.202924547+01:00 level=info msg="Delete service account" service=grafana-appwithserviceaccount-app orgID=1 saID=16
  logger=serviceauth.extsvcaccounts t=2023-11-14T17:02:35.27970295+01:00 level=info msg="Deleted external service account" service=grafana-appwithserviceaccount-app orgID=1
  ```
- Remove the other plugin:
  ```bash
  ./grafana/bin/grafana-cli plugins remove grafana-appwithonbehalfofauth-app
  # Removing plugin: grafana-appwithonbehalfofauth-app
  # Please restart Grafana after installing or removing plugins. Refer to Grafana documentation for instructions if necessary.
  ```
- Restart grafana and check that the service account for `grafana-appwithserviceaccount-app` has been deleted.
- You should also see the following logs:
  ```log
  logger=extsvcauth.registry t=2023-11-14T16:54:41.544612593+01:00 level=info msg="Detected removed External Service" service=grafana-appwithonbehalfofauth-app provider=OAuth2Server
  logger=oauthserver t=2023-11-14T16:54:41.544699244+01:00 level=info msg="Remove external service" service=grafana-appwithonbehalfofauth-app
  logger=serviceauth.extsvcaccounts t=2023-11-14T16:54:42.554682225+01:00 level=info msg="Delete service account" service=grafana-appwithonbehalfofauth-app orgID=1 saID=17
  logger=serviceauth.extsvcaccounts t=2023-11-14T16:54:42.565611239+01:00 level=info msg="Deleted external service account" service=grafana-appwithonbehalfofauth-app orgID=1
  ```
  

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Review by andresmgot — COMMENTED** [review]
Sounds good to me, I haven't tried this time though. One question, is this "thread" safe? Meaning, is it possible that the job that checks if the service has been registered this time runs before the normal loading system has finished?

**Comment by gamab** [comment]
Good question. 
**TL; DR;** It cannot happen. Registration happens on initialization, cleanup happens on run.

Here is the callstack when reaching the plugin registration:
```
serviceregistration.(*Service).RegisterExternalService (~/grafana/grafana/pkg/services/pluginsintegration/serviceregistration/serviceregistration.go:54)
pipeline.(*ExternalServiceRegistration).Register (~/grafana/grafana/pkg/services/pluginsintegration/pipeline/steps.go:43)
pipeline.(*ExternalServiceRegistration).Register-fm (Unknown Source:1)
initialization.(*Initialize).Initialize (~/grafana/grafana/pkg/plugins/manager/pipeline/initialization/initialization.go:54)
loader.(*Loader).Load (~/grafana/grafana/pkg/plugins/manager/loader/loader.go:57)
loader.(*Loader).Load (~/grafana/grafana/pkg/services/pluginsintegration/loader/loader.go:30)
pluginstore.ProvideService (~/grafana/grafana/pkg/services/pluginsintegration/pluginstore/store.go:40)
server.Initialize (~/grafana/grafana/pkg/server/wire_gen.go:377)
commands.RunServer (~/grafana/grafana/pkg/cmd/grafana-server/commands/cli.go:114)
commands.ServerCommand.func1 (~/grafana/grafana/pkg/cmd/grafana-server/commands/cli.go:41)
v2.(*Command).Run (~/go/pkg/mod/github.com/urfave/cli/v2@v2.25.0/command.go:273)
v2.(*Command).Run (~/go/pkg/mod/github.com/urfave/cli/v2@v2.25.0/command.go:266)
v2.(*App).RunContext (~/go/pkg/mod/github.com/urfave/cli/v2@v2.25.0/app.go:332)
v2.(*App).Run (~/go/pkg/mod/github.com/urfave/cli/v2@v2.25.0/app.go:309)
main.main (~/grafana/grafana/pkg/cmd/grafana/main.go:53)
runtime.main (/usr/local/go/src/runtime/proc.go:267)
```

Plugin loading happens during the `server.Initialize` phase:
https://github.com/grafana/grafana/blob/2a522241cc5f1c24589e9f0a6ef1e0a1b7ede484/pkg/cmd/grafana-server/commands/cli.go#L114

Background jobs are started later with `s.Run`:
https://github.com/grafana/grafana/blob/2a522241cc5f1c24589e9f0a6ef1e0a1b7ede484/pkg/cmd/grafana-server/commands/cli.go#L130

Please do not hesitate to tell me if I missed something :)

**Comment by andresmgot** [comment]
Cool, that's right, thanks for confirming :+1:

**Review by Jguer — APPROVED** [review]
